src/unix/core.c

```diff
@@ -983,3 +987,47 @@ int uv__dup2_cloexec(int oldfd, int newfd) {
     return r;
   }
 }
+
+
+int uv_homedir(char* home, size_t* len) {
+
+  char login[LOGIN_NAME_MAX];
+  struct passwd pw;
+  struct passwd* result;
+  char* buf;
+  size_t loginsize = sizeof(login);
+  size_t bufsize = sysconf(_SC_GETPW_R_SIZE_MAX);
+  int r;
+
+  if (home == NULL || len == NULL || *len == 0)
+    return -EINVAL;
+
+  r = getlogin_r(login, loginsize);
+
+  if (r != 0)
+    return -r;
+
+  buf = malloc(bufsize);
+
+  if (buf == NULL)
+    return -ENOMEM;
+
+  r = getpwnam_r((const char*) login, &pw, buf, bufsize, &result);
+
+  if (r != 0) {
+    free(buf);
+    return -r;
+  }
+
+  if (result == NULL) {
+    free(buf);
+    return -ENOENT;
+  }
+
+  strncpy(home, (const char*) pw.pw_dir, *len);
+  free(buf);
+  home[*len - 1] = '\0';
+  *len = strlen(home);
```

Comment:
I would `strlen()` first, clamp the result, then `memcpy()` and add the zero byte if necessary.

`strncpy()` + `strlen()` is not wrong but `strncpy()` also unnecessarily zeros the part of the buffer after the string.